build.gradle

```diff
@@ -212,6 +212,7 @@ project(':iceberg-core') {
       exclude group: 'org.tukaani' // xz compression is not supported
     }
 
+    implementation 'org.apache.httpcomponents.client5:httpclient5'
```

Comment:
judging by https://github.com/square/okhttp/issues/3472 and https://www.mocklab.io/blog/which-java-http-client-should-i-use-in-2020/ I wonder if it would make sense to use OkHttp rather than the Apache HttpClient stuff

Reply:
Yes I do tend to agree that the OkHTTP library is more feature rich and intuitive. There's also a large amount of code we can repurpose from Trino etc with it.

I think since this is already written, to unblock integrating the actual catalog in, it might be good to add this implentation so that other tasks can be worked on, and then possibly refactor this class after that if we decide to.

Reply:
I checked the runtime classpath by adding OkHTTP 4.9.3 to `iceberg-core`, and it brings in a very small set of dependencies. One of which is the kotlin stdlib, but people seem to use it without worrying about kotlin interoperability issues quite often.

```
➜  iceberg git:(master) ✗ ./gradlew iceberg-core:dependencies --configuration=runtimeClasspath

> Task :iceberg-core:dependencies

------------------------------------------------------------
Project ':iceberg-core'
------------------------------------------------------------

runtimeClasspath - Runtime classpath of source set 'main'.
+--- project :iceberg-api
|    +--- org.slf4j:slf4j-api -> 1.7.30
|    +--- com.github.stephenc.findbugs:findbugs-annotations -> 1.3.9-1
|    \--- project :iceberg-bundled-guava
+--- org.slf4j:slf4j-api -> 1.7.30
+--- com.github.stephenc.findbugs:findbugs-annotations -> 1.3.9-1
+--- project :iceberg-common
|    +--- org.slf4j:slf4j-api -> 1.7.30
|    +--- com.github.stephenc.findbugs:findbugs-annotations -> 1.3.9-1
|    \--- project :iceberg-bundled-guava
+--- project :iceberg-bundled-guava
+--- org.apache.avro:avro -> 1.10.1
|    +--- com.fasterxml.jackson.core:jackson-core:2.11.3 -> 2.11.4
|    +--- com.fasterxml.jackson.core:jackson-databind:2.11.3 -> 2.11.4
|    |    +--- com.fasterxml.jackson.core:jackson-annotations:2.11.4
|    |    \--- com.fasterxml.jackson.core:jackson-core:2.11.4
|    +--- org.apache.commons:commons-compress:1.20
|    \--- org.slf4j:slf4j-api:1.7.30
+--- com.fasterxml.jackson.core:jackson-databind -> 2.11.4 (*)
+--- com.fasterxml.jackson.core:jackson-core -> 2.11.4
+--- com.github.ben-manes.caffeine:caffeine -> 2.8.4
|    +--- org.checkerframework:checker-qual:3.4.0
|    \--- com.google.errorprone:error_prone_annotations:2.3.4
+--- org.roaringbitmap:RoaringBitmap -> 0.9.22
|    \--- org.roaringbitmap:shims:0.9.22
\--- com.squareup.okhttp3:okhttp:4.9.3
     +--- com.squareup.okio:okio:2.8.0
     |    +--- org.jetbrains.kotlin:kotlin-stdlib:1.4.0 -> 1.4.10
     |    |    +--- org.jetbrains.kotlin:kotlin-stdlib-common:1.4.10
     |    |    \--- org.jetbrains:annotations:13.0
     |    \--- org.jetbrains.kotlin:kotlin-stdlib-common:1.4.0 -> 1.4.10
     \--- org.jetbrains.kotlin:kotlin-stdlib:1.4.10 (*)
```

And without

```
➜  iceberg git:(master) ✗ ./gradlew iceberg-core:dependencies --configuration=runtimeClasspath

> Task :iceberg-core:dependencies

------------------------------------------------------------
Project ':iceberg-core'
------------------------------------------------------------

runtimeClasspath - Runtime classpath of source set 'main'.
+--- project :iceberg-api
|    +--- org.slf4j:slf4j-api -> 1.7.30
|    +--- com.github.stephenc.findbugs:findbugs-annotations -> 1.3.9-1
|    \--- project :iceberg-bundled-guava
+--- org.slf4j:slf4j-api -> 1.7.30
+--- com.github.stephenc.findbugs:findbugs-annotations -> 1.3.9-1
+--- project :iceberg-common
|    +--- org.slf4j:slf4j-api -> 1.7.30
|    +--- com.github.stephenc.findbugs:findbugs-annotations -> 1.3.9-1
|    \--- project :iceberg-bundled-guava
+--- project :iceberg-bundled-guava
+--- org.apache.avro:avro -> 1.10.1
|    +--- com.fasterxml.jackson.core:jackson-core:2.11.3 -> 2.11.4
|    +--- com.fasterxml.jackson.core:jackson-databind:2.11.3 -> 2.11.4
|    |    +--- com.fasterxml.jackson.core:jackson-annotations:2.11.4
|    |    \--- com.fasterxml.jackson.core:jackson-core:2.11.4
|    +--- org.apache.commons:commons-compress:1.20
|    \--- org.slf4j:slf4j-api:1.7.30
+--- com.fasterxml.jackson.core:jackson-databind -> 2.11.4 (*)
+--- com.fasterxml.jackson.core:jackson-core -> 2.11.4
+--- com.github.ben-manes.caffeine:caffeine -> 2.8.4
|    +--- org.checkerframework:checker-qual:3.4.0
|    \--- com.google.errorprone:error_prone_annotations:2.3.4
\--- org.roaringbitmap:RoaringBitmap -> 0.9.22
     \--- org.roaringbitmap:shims:0.9.22
```
(*) - dependencies omitted (listed previously)
```

Reply:
It looks like the v4 of OkHttp is what brings in Kotlin: https://github.com/square/okhttp/issues/472